src/main/java/euteam/cleanmachine/model/facility/Machine.java

```diff
@@ -1,83 +1,138 @@
+        if(runningState.getEndTime()<= System.currentTimeMillis()){
+            lockMachine(runningState.getUserId());
+            return;
+        }
     }
-  
-    public void startMachine(Long userId){
-        state.startMachine(this,userId);
+
+    public boolean idle(){
+        try {
+            state.idle(this);
+        }catch (Exception e){
+            return  false ;
+        }
+        return true;
     }
-  
     public void authenticateOnMachine(Long userId){
         state.authenticateOnMachine(this,userId);
     }
-  
+    public boolean startMachine(Long userId, long programId){
+        Program p  =  getProgramById(programId);
+        long durationInMiliseconds =Math.round(p.getDuration() * 60 * 1000);
+        try {
+            state.startMachine(this, userId, durationInMiliseconds + System.currentTimeMillis(), programId);
+        }catch (StateTransitionException e){
+            return false;
+        }
+        return true;
+    }
+    private void lockMachine(Long userId){
+        state.lockMachine(this,userId);
+    }
+    public boolean unlockMachine(long userId){
+        Long loggedInUserID = null;
+        if(state instanceof LockedState){
+            LockedState state = (LockedState) getState();
+          loggedInUserID=  state.getUserId();
+        }else if(state instanceof RunningState){
+            RunningState state = (RunningState) getState();
+            loggedInUserID=  state.getUserId();
+        }
+        if(loggedInUserID == userId){
+            idle();
+            return true;
+        }
+        return false;
+    }
+    public Long getProgramEndTime(){
+        if( state instanceof  RunningState) {
+            RunningState runningState = (RunningState) state;
+            return runningState.getEndTime();
+        }
+        return null;
+    }
     public void outOfOrder(Long employeId){
         state.outOfOrder(this,employeId);
     }
-  
-    public void reopenMachine(Long employeId){
-        state.reOpenMachine(this);
-    }
 
-    public Long getId() {
-        return id;
-    }
 
-    public void setId(Long id) {
-        this.id = id;
+    public Long getLoggedInUserId(){
+        try {
+            AuthenticatedState authenticatedState = (AuthenticatedState) state;
+            return authenticatedState.getUserId();
+        }catch (Exception e){
+            return null;
+        }
+
+    }
+    public MachineState getState() {
+        return state;
     }
 
     public String getIdentifier() {
         return identifier;
     }
-
     public void setIdentifier(String identifier) {
         this.identifier = identifier;
     }
 
     public List<Program> getPrograms() {
         return programs;
     }
-
     public void setPrograms(List<Program> programs) {
         this.programs = programs;
     }
+
+    public boolean containsProgram(Long programId){
+        for (Program program:programs
+             ) {
+            if(program.getId()==programId)return true;
```

Comment:
Looks like when you try to start the woll program through the web interface it tells that the program doesn't exist, so this change fixes the bug
```suggestion
            if(program.getId().equals(programId)) return true;
```